From b142da8d2c9e2e2dfbe4e89e680dd124f6064ac8 Mon Sep 17 00:00:00 2001
From: Pablo Neira Ayuso <pablo@netfilter.org>
Date: Mon, 3 Feb 2014 12:09:29 +0100
Subject: [PATCH 4/6] libnfnetlink: initialize attribute padding to resolve
 valgrind warnings

==12195== Syscall param socketcall.sendto(msg) points to uninitialised byte(s)
==12195==    at 0x51209C3: __sendto_nocancel (syscall-template.S:81)
==12195==    by 0x53E4D12: nfnl_send (libnfnetlink.c:391)
==12195==    by 0x53E6952: nfnl_query (libnfnetlink.c:1569)
==12195==    by 0x4E344AF: __build_send_cfg_msg.isra.1 (libnetfilter_log.c:143)
==12195==    by 0x4E34710: nflog_bind_group (libnetfilter_log.c:413)
==12195==    by 0x400CB1: main (nfulnl_test.c:77)
==12195==  Address 0x7fefff3e9 is on thread 1's stack

This patch sets to zero the padding that is included to align the
attribute payload.

Reported-by: Ivan Homoliak <xhomol11@gmail.com>
Signed-off-by: Pablo Neira Ayuso <pablo@netfilter.org>
---
 src/libnfnetlink.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/src/libnfnetlink.c b/src/libnfnetlink.c
index 4b2bcd0..398b7d7 100644
--- a/src/libnfnetlink.c
+++ b/src/libnfnetlink.c
@@ -809,6 +809,7 @@ int nfnl_addattr_l(struct nlmsghdr *n, int maxlen, int type, const void *data,
 	nfa->nfa_type = type;
 	nfa->nfa_len = len;
 	memcpy(NFA_DATA(nfa), data, alen);
+	memset((uint8_t *)nfa + nfa->nfa_len, 0, NFA_ALIGN(alen) - alen);
 	n->nlmsg_len = (NLMSG_ALIGN(n->nlmsg_len) + NFA_ALIGN(len));
 	return 0;
 }
-- 
2.12.1

